src/Microsoft.Restier.WebApi/Query/QueryExecutorWebApi.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Web.OData.Extensions;
+using Microsoft.Restier.Core.Query;
+
+namespace Microsoft.Restier.WebApi.Query
+{
+    internal class QueryExecutorWebApi : IQueryExecutor
+    {
+        public IQueryExecutor Inner { get; set; }
+
+        public async Task<QueryResult> ExecuteQueryAsync<TElement>(
+            QueryContext context,
+            IQueryable<TElement> query,
+            CancellationToken cancellationToken)
+        {
+            var queryContext = context.ApiContext.GetApiService<WebApiContext>();
+            if (queryContext.QueryIncludeTotalCount == true)
+            {
+                var countQuery = ExpressionHelpers.GetCountableQuery(query);
+                var expression = countQuery.Expression;
+                expression = ExpressionHelpers.Count(expression, countQuery.ElementType);
+                var result = await ExecuteSingleAsync<long>(context, countQuery, expression, cancellationToken);
```

Comment:
General comment: I checked the code, seems ExecuteSingleAsync key logic will be countQuery.provider.execute(expression), then question will be whether we need two parameters here (query and expression) or rename query to QueryProvider?